sys/aarch64/boot.c

```diff
@@ -276,50 +262,50 @@ __boot_text static void enable_mmu(pde_t *pde) {
 __boot_text __noreturn void aarch64_init(paddr_t dtb) {
   drop_to_el1();
   configure_cpu();
+  boot_clear(PHYSADDR(__bss), PHYSADDR(__ebss));
+  boot_sbrk_init(PHYSADDR(__ebss));
+
+  vaddr_t dtb_va = VIRTADDR(boot_save_dtb(dtb));
+
+  /* Make sure DTB is mapped into kernel virtual address space. */
+  vaddr_t kernel_end = VIRTADDR(boot_sbrk_align(PAGESIZE));
 
-  /* Set end address of kernel for boot allocation purposes. */
-  _bootmem_end = (void *)align(PHYSADDR(__ebss), PAGESIZE);
+  pde_t *pde = build_page_table(kernel_end);
+  enable_mmu(pde);
 
-  pde_t *kernel_pde = build_page_table();
-  enable_mmu(kernel_pde);
+  _bootmem_end = boot_sbrk(0);
 
   vaddr_t boot_sp = (vaddr_t)&_boot_stack[PAGESIZE];
 
   __asm __volatile("mov x0, %0\n\t"
                    "mov x1, %1\n\t"
-                   "mov sp, %2\n\t"
-                   "br %3"
+                   "mov x2, %2\n\t"
+                   "mov sp, %3\n\t"
+                   "br %4"
                    :
-                   : "r"(dtb), "r"(kernel_pde), "r"(boot_sp), "r"(aarch64_boot)
-                   : "x0", "x1");
+                   : "r"(dtb_va), "r"(pde), "r"(kernel_end), "r"(boot_sp),
+                     "r"(aarch64_boot)
+                   : "x0", "x1", "x2");
   __unreachable();
 }
 
-static void clear_bss(void) {
-  uint64_t *ptr = (uint64_t *)__bss;
-  uint64_t *end = (uint64_t *)__ebss;
-  while (ptr < end)
-    *ptr++ = 0;
-}
-
 extern void *board_stack(void);
 
-static __noreturn void aarch64_boot(paddr_t dtb, paddr_t pde) {
-  clear_bss();
+static __noreturn void aarch64_boot(void *dtb, paddr_t pde,
+                                    vaddr_t kernel_end) {
+  vm_kernel_end = kernel_end;
 
 #if KASAN
   _kasan_sanitized_end =
     KASAN_SANITIZED_START + BOOT_KASAN_SANITIZED_SIZE((vaddr_t)__ebss);
 #endif
 
-  pmap_bootstrap(pde, (pde_t *)(pde + DMAP_BASE));
-
-  FDT_early_init(dtb, phys_to_dmap(dtb));
-  void *fdtp = phys_to_dmap(FDT_get_physaddr());
-  FDT_changeroot(fdtp);
+  FDT_init(dtb);
 
   void *sp = board_stack();
 
+  pmap_bootstrap(pde, (pde_t *)(pde + DMAP_BASE));
```

Comment:
I don't understand why this has been moved after the call to `board_stack`. I know that RISCV does so, but it's only because it relies on the kevn. `pmap_bootstrap` should be called as soon as possible (as it should construct the dmap).

Reply:
To make it look exactly as `riscv_boot` procedure. I'm going to address a mess with dmap in future PRs.